### ISSUE TITLE
Current state with master branch, GR_jll v0.69.1+0 (v0.69.1+1 yanked)

### ISSUE THREAD

**Issue by mkitti** [jheinen/GR.jl#490]
Because there may be some confusion, I wanted to clarify the current state of affairs.

The master branch currently has #483 merged. This adds several capabilities:
1. It now contains portions of the old build script as `GR.GRPreferences.Downloader`. The main entry point there is `GR.GRPreferences.Downloader.download() ` which should help to download and extract the sciapp/gr tarball.
2. We now have a `GR.GRPreferences.use_upstream_binary()`which invokes the above `download` function and sets GR to use that via `GR.GRPreferences.use_system_binary()`.
3. `GR.GRPreferences` now also installs overrides for `GR_jll`. There is a `depot` based mechanism for doing so via `Overrides.toml` and a `LocalPreferences.toml` based mechanism that can be more local to the project. The default is to use the `depot` based mechanism.
4. It is compatible with the GR_jll v0.69.1+1 `dont_dlopen` patch due to https://github.com/mkitti/GR.jl/blob/005b721753beb8e2af7dc6d2eb459128ad3fe466/src/preferences.jl#L59-L65

Since GR_jll v0.69.1+1 was yanked:
1. `LD_PRELOAD` may be needed without first invoking `GR.GRPreferences.use_system_binary()` and restarting Julia. This is because `GR_jll` is loaded automatically and it loads the BinaryBuilder.org binaries regardless of `GRDIR` or the `GRPreferences` settings.
2. For a system install of GR_jll, I highly recommend the use of an Overrides.toml within the system depot artifacts directory.
3. I believe we should repropose GR_jll v0.69.1+1 as GR_jll v0.70.0+0. This will stop GR_jll from loading its binaries automatically.

**Comment by mkitti** [comment]
To clarify, I mean the master branch at [da5cfd3](https://github.com/jheinen/GR.jl/commit/da5cfd32674153473f725e5ce893e3f7dc112a71)

**Comment by jheinen** [comment]
I'll make a new attempt and release a new `GR` run-time release (from the current development tree) and `GR.jl` release soon (0.71.0). I hope, the `Plots` problems will be solved with the update.

**Comment by PaulSoderlind** [comment]
today (24 Nov) after `up` (which updated GR to 0.71) on a Win64 machine:
```
connect: No error
GKS: can't connect to GKS socket application
...
```

**Comment by jheinen** [comment]
For any reason, `libcrypto-1_1-x64.dll` and `libssl-1_1-x64.dll` are not found. I'm pretty shure it worked in my interactive tests. We even tested with Unicode paths. I suspect some deps in the Qt5 libraries.

**Comment by PaulSoderlind** [comment]
did that, restarted Julia, did a `build GR` and the tried plotting. Same error message. (Win11)

**Comment by jheinen** [comment]
Could you please `set JULIA_DEBUG=GR` and then start Julia?

**Comment by jheinen** [comment]
@mkitti : It seems, that we should not escape the path delimiters (`\\` => `\`) and the `gksqt` command (using `"`). Without the double path separators and the surrounding `"` (in the `gksqt` command) it seems to work. I have absolutely no idea, why this worked in my tests before.

**Comment by jheinen** [comment]
It's no longer necessary to set the Windows PATH when starting `gksqt.exe` as the DLL directories (for the required artifacts) are already added before (using `AddDllDirectory`).

@PaulSoderlind : Could you please switch back to the JLL version
```
using GR
GR.GRPreferences.use_jll_binary(; force = true)
```
 and try `GR#master`. Sorry for the noise and thanks in advance.

**Comment by PaulSoderlind** [comment]
tried. Got a pop-up window saying "gksqt.exe - Entry Point Not Found...." and then eventually the usual 
```connect: No error
GKS: can't connect to GKS socket application```

**Comment by jheinen** [comment]
@mkitti , @t-bltg : As already suspected, the environment variable for `GKS_QT` is (internally) truncated. We will have to provide a patch in the short term and create a new `GR_jll` version (0.71.1). The build step for `GR v0.71.1` has already been triggered. Affected are only Windows users who want to use the BB artifacts.

**Comment by jheinen** [comment]
@t-bltg : I can confirm that `GR 0.71.1` now works on Windows. As soon as `GR.jl` is tagged, we should bump GR in `Plots.jl`.

**Comment by mkitti** [comment]
> @mkitti : It seems, that we should not escape the path delimiters (`\\` => `\`) and the `gksqt` command (using `"`). Without the double path separators and the surrounding `"` (in the `gksqt` command) it seems to work. I have absolutely no idea, why this worked in my tests before.

I'm not quite following, but it does not seem you actually changed anything in GR.jl to resolve this issue, and that you found that the problem was with the environment variable being truncated in the gr framework binaries.

> It's no longer necessary to set the Windows PATH when starting `gksqt.exe` as the DLL directories (for the required artifacts) are already added before (using `AddDllDirectory`).

I'm not sure if this applies to gksqt.exe since we are starting a new process, and we are not using `SetDllDirectoryW`  or even `SetDefaultDllDirectories` at the moment. We are only using `AddDllDirectory`.

`SetDllDirectoryW` does affect child processes, but I do not think `SetDefaultDllDirectories` or `AddDllDirectory` does. From the [`SetDefaultDllDirectories`](https://learn.microsoft.com/en-us/windows/win32/api/libloaderapi/nf-libloaderapi-setdefaultdlldirectories) documentation:
>  The process DLL search path applies only to the calling process and persists for the life of the process.
Comapre that 

Compare that to the explicit note for [`SetDllDirectoryW`](https://learn.microsoft.com/en-us/windows/win32/api/winbase/nf-winbase-setdlldirectoryw):

> For Win32 processes that are not running a packaged or protected process, calling this function will also affect the DLL search order of the children processes started from the process that has called the function.

**Comment by mkitti** [comment]
> OK, it now works. After rebooting the PC. Anyhow the message is now

I'm not sure why rebooting the PC was necessary. Restarting Julia should have been sufficient. My one theory is that we had a rogue gksqt.exe process running.

It does seem that the following was helpful in at least temporarily resolving the issue though.

```
using GR
GR.GRPreferences.use_upstream_binary(; force = true)
```

**Comment by mkitti** [comment]
Since all the issues raised in the original note have been resolved, I am closing this issue.

Specifically,

> LD_PRELOAD may be needed without first invoking GR.GRPreferences.use_system_binary() and restarting Julia. This is because GR_jll is loaded automatically and it loads the BinaryBuilder.org binaries regardless of GRDIR or the GRPreferences settings.

This was addressed with `dont_dlopen` being used in GR_jll and being released in the v0.70+ series of JLL binaries:
https://github.com/JuliaPackaging/Yggdrasil/blob/8000e3e7e9880c10a265ef37c8ca7f7383d232a4/G/GR/build_tarballs.jl#L76-L79
